formats/protobuf/src/main/java/io/cloudevents/protobuf/ProtoDeserializer.java

```diff
@@ -0,0 +1,107 @@
+
+    public ProtoDeserializer(CloudEvent protoCe) {
+        this.protoCe = protoCe;
+    }
+
+    @Override
+    public <W extends CloudEventWriter<R>, R> R read(
+        CloudEventWriterFactory<W, R> writerFactory,
+        CloudEventDataMapper<? extends CloudEventData> mapper) throws CloudEventRWException {
+        SpecVersion specVersion = SpecVersion.parse(this.protoCe.getSpecVersion());
+        if (specVersion != SpecVersion.V1) {
+            // The protospec is only defined for V1 so far.
+            throw CloudEventRWException.newInvalidSpecVersion(specVersion.toString());
+        }
+
+        final CloudEventWriter<R> writer = writerFactory.create(specVersion);
+
+        // Required attributes
+        writer.withContextAttribute(CloudEventV1.ID, this.protoCe.getId());
+        writer.withContextAttribute(CloudEventV1.SOURCE, this.protoCe.getSource());
+        writer.withContextAttribute(CloudEventV1.TYPE, this.protoCe.getType());
+
+        // Optional attributes
+        for (Entry<String, CloudEventAttributeValue> entry : this.protoCe.getAttributesMap().entrySet()) {
+            String name = entry.getKey();
+            CloudEventAttributeValue val = entry.getValue();
+            switch (val.getAttrCase()) {
+                case CE_BOOLEAN:
+                    writer.withContextAttribute(name, val.getCeBoolean());
+                    break;
+                case CE_INTEGER:
+                    writer.withContextAttribute(name, val.getCeInteger());
+                    break;
+                case CE_STRING:
+                    writer.withContextAttribute(name, val.getCeString());
+                    break;
+                case CE_BYTES:
+                    final byte[] rawBytes = val.getCeBytes().toByteArray();
+                    final String base64String = Base64.getEncoder().encodeToString(rawBytes);
+                    writer.withContextAttribute(name, base64String);
+                    break;
+                case CE_URI:
+                    writer.withContextAttribute(name, URI.create(val.getCeUri()));
+                    break;
+                case CE_URI_REF:
+                    writer.withContextAttribute(name, URI.create(val.getCeUriRef()));
+                    break;
+                case CE_TIMESTAMP:
+                    com.google.protobuf.Timestamp timestamp = val.getCeTimestamp();
+                    Instant instant = Instant.ofEpochSecond(timestamp.getSeconds(), timestamp.getNanos());
+                    writer.withContextAttribute(name, instant.atOffset(ZoneOffset.UTC));
+                    break;
+                case ATTR_NOT_SET:
+                    // I'm not sure if an unset attribute is a valid usecase, but I am leaning towards no. If this was JSON,
+                    // I would assign "null" here, but I don't see a way of doing that in the Java SDK.
+                    throw CloudEventRWException.newInvalidAttributeValue(name, val, null);
```

Comment:
There is something here I'm missing about the protobuf spec i guess... @JemDay can you check out this one?

Reply:
In this case, someone constructed one of these CloudEventAttribute objects, but didn't actually set the value within it. It is unclear to me if CloudEvents allows for setting an attribute with no value. I feel like the answer is no, since attributes must have a type and an empty attribute's type is ambiguous.

The JSON format allows for null to be an attribute value and says to treat it as "unset or omitted". I think treating this case as omitted makes sense, but it might be worth codifying that.

Reply:
But in this case you're not threating as omitted, you're throwing an exception, failing the conversion right?

Reply:
I removed the exception and chose to just do nothing to treat this as omitted.